os/se/ameba/security_ameba_wrapper.c

```diff
@@ -0,0 +1,1933 @@
+	uint32_t aes_key_len;
+
+	rtl_rsa_context rsa_ctx;
+	rtl_ecdh_context ecdh_ctx;
+	rtl_ecdsa_context ecdsa_ctx;
+	rtl_dhm_context dhm_ctx;
+
+	unsigned char hmac_key[32];
+
+	unsigned char *cert_data;
+	uint32_t cert_len;
+} key_struc;
+
+static key_struc key_storage[KEY_STORAGE_INDEX_MAX] = { 0 };
+static uint32_t init_stat = HAL_NOT_INITIALIZED;
+
+#define HAL_RTL_CHECKBUSY() while (init_stat == HAL_BUSY) {}
+
+/* Extern api from rtl_security_api_wrapper.c */
+extern uint32_t rtl_cryptoEngine_init_wrapper(void);
+extern uint32_t rtl_write_storage_wrapper(uint32_t ss_idx, hal_data *data);
+extern uint32_t rtl_read_storage_wrapper(uint32_t ss_idx, hal_data *data);
+extern uint32_t rtl_delete_storage_wrapper(uint32_t ss_idx);
+extern uint32_t rtl_cryptoAES_ecb_wrapper(uint8_t* key, uint32_t keylen, unsigned char* message, uint32_t msglen, unsigned char* pResult, uint8_t mode);
+extern uint32_t rtl_cryptoAES_cbc_wrapper(uint8_t* key, uint32_t keylen, unsigned char* message, uint32_t msglen, unsigned char* pResult, uint8_t mode);
+extern uint32_t rtl_cryptoAES_ctr_wrapper(uint8_t* key, uint32_t keylen, unsigned char* message, uint32_t msglen, unsigned char* pResult, uint8_t mode);
+
+/**
+ * Helper function
+ */
+static void* my_calloc(size_t nelements, size_t elementSize)
+{
+	size_t size;
+	void *ptr = NULL;
+
+	size = nelements * elementSize;
+	ptr = malloc(size);
+
+	if (ptr)
+		memset(ptr, 0, size);
+
+	return ptr;
+}
+
+/* Set prime modulus and generator */
+int rtl_dhm_set_group(rtl_dhm_context *ctx, const rtl_mpi *P, const rtl_mpi *G)
+{
+	int ret;
+
+	if (ctx == NULL || P == NULL || G == NULL) {
+		return HAL_FAIL;
+	}
+
+	if ((ret = rtl_mpi_copy(&ctx->P, P)) != 0 || (ret = rtl_mpi_copy(&ctx->G, G)) != 0) {
+		return HAL_FAIL;
+	}
+
+	ctx->len = rtl_mpi_size(&ctx->P);
+	return 0;
+}
+
+static int hal_rand(void *rng_state, unsigned char *output, size_t len)
+{
+	rtw_get_random_bytes(output, len);
+	return HAL_SUCCESS;
+}
+
+/**
+ * Common
+ */
+int ameba_hal_init(hal_init_param *params)
+{
+	AWRAP_ENTER;
+
+	uint32_t ret = HAL_SUCCESS;
+
+	if (init_stat != HAL_NOT_INITIALIZED) {
+		return init_stat;
+	}
+
+	for (uint32_t i = 0; i < KEY_STORAGE_INDEX_MAX; i++) {
+		memset(key_storage[i].aes_key, 0, sizeof(key_storage[i].aes_key));
+		key_storage[i].aes_key_len = 0;
+
+		rtl_rsa_init(&(key_storage[i].rsa_ctx), RTL_RSA_PKCS_V15, 0);
+		rtl_ecdh_init(&(key_storage[i].ecdh_ctx));
+		rtl_ecdsa_init(&(key_storage[i].ecdsa_ctx));
+		rtl_dhm_init(&(key_storage[i].dhm_ctx));
+
+		memset(key_storage[i].hmac_key, 0, sizeof(key_storage[i].hmac_key));
+	}
+
+	/* Realtek added to initialize ROM code calloc & free function handler */
+	p_rom_ssl_ram_map = (struct _rom_mbedtls_ram_map*)&rom_ssl_ram_map;
+	p_rom_ssl_ram_map->ssl_calloc = my_calloc;
+	p_rom_ssl_ram_map->ssl_free = free;
+
+	ret = rtl_cryptoEngine_init_wrapper();					/* init crypto engine */
+	if (ret == HAL_SUCCESS) {
+		init_stat = HAL_SUCCESS;
```

Comment:
rtl_cryptoEngine_init_wrapper() doesn't return RTL_HAL_SUCCESS not HAL_SUCCESS.

Reply:
RTL_HAL_SUCCESS is same as HAL_SUCCESS, RTL_HAL_SUCCESS is a local define to fix duplicate hal enum problem.
In TizenRT HAL enum (Security_hal.h) there are some same name enum as RTL. We cannot include Security_hal.h, but we will need to return the value, so we using local define to return the result.

Reply:
I know that RTL_HAL_SUCCESS and HAL_SUCCESS are same. this code make readability bad. so please modify all functions that return wrong type for consistency

Reply:
understood, previously we are trying to make the return code same as defined enum on security_hal.h.
So in this case we will change the return code of the function to 1 for fail and 0 for success. Is that acceptable?, if it is acceptable then we will apply the change.

Reply:
Instead of using 1, 0(it reduces readability) how about below?
	ret = rtl_cryptoEngine_init_wrapper();					/* init crypto engine */
	if (ret == RTL_HAL_SUCCESS) {
		init_stat = HAL_SUCCESS;
	}

Reply:
For all the rtl api return we have change it to what you have suggested.